issue_5/contracts/GreenVault.sol

```diff
@@ -1,93 +1,172 @@
 // SPDX-License-Identifier: MIT
 pragma solidity >=0.8.4;
 
+import "./external/chainlink/AggregatorV3Interface.sol";
+
 import "@yield-protocol/utils-v2/contracts/access/Ownable.sol";
 import "@yield-protocol/utils-v2/contracts/token/ERC20.sol";
 
 /**
 @dev Internal library that implements Fixed Point Math
-All numbers are WADs - fixed point numbers with decimals=18
+WADs are fixed point numbers with decimals=18
  */
 library FPMath {
-    uint256 internal constant WAD = 1e18;
-
     /**
     @notice Multiply 2 WADs
     */
     function wadMul(uint256 a, uint256 b) internal pure returns (uint256) {
+        return mul(a, b, 18);
+    }
+
+    /**
+    @notice Multiply 2 fixed point numbers
+    */
+    function mul(
+        uint256 a,
+        uint256 b,
+        uint8 bDecimals
+    ) internal pure returns (uint256) {
         uint256 ret = (a * b);
-        require(b == 0 || ret / b == a, "wadMul: overflow");
-        return ret / WAD;
+        // require(b == 0 || ret / b == a, "mul: overflow");
+        // ^ not needed since solidity 0.8
+        return ret / (10**bDecimals);
     }
 }
 
 /**
 @title GreenVault
-@notice A (single-token) vault that mints wrapped tokens
-in response to deposits
+@notice A (single-token) vault that allows users to deposit ETH and borrow DAI
  */
-contract GreenVault is ERC20("GreenVaultToken", "OWL_GVT", 18), Ownable {
+contract GreenVault is Ownable {
     /**
     @notice the only token this vault supports
      */
     IERC20Metadata public theOnlySupportedToken;
 
     /**
-    Multiplier: for each token deposited we mint `exchangeRateWad` Vault Tokens
+    @dev price feed
+    */
+    AggregatorV3Interface immutable priceFeed;
+
+    /**
+    @dev per-user data
+     */
+    struct UserData {
+        /**
+        @dev how much ETH was deposited
+         */
+        uint256 deposited;
+        /**
+        @dev how much ETH was borrowed against
+         */
+        uint256 borrowed;
```

Comment:
I thought that this was a neat implementation initially. By tracking deposited and borrowed both in ETH terms, you don't need fixed point division in `payback`, and you have no risk of considering rounding.

However, it doesn't work because of liquidation mechanics. Liquidation happens when the system thinks that the DAI it lent won't be recovered because of changing oracle prices.

I deposit 1 ETH at 3000 ETH/DAI and get 3000 DAI.
The DAI/ETH price drops to 2500 ETH/DAI, the user has 3000 DAI, he's not going to repay that to get 1 ETH (worth 2500 DAI).

That is usually solved by overcollateralization. Yield v1 (and MakerDAO) used 150%. Meaning you need to deposit 1.5 ETH at 3000 ETH/DAI to borrow 3000 DAI.

You will be liquidated if your collateralization ratio falls below 150%, meaning that the value of the 1.5 ETH you deposited falls below 4500 DAI. That 50% difference is to give time to the liquidation engine to sell your collateral in a scenario where prices might be dropping.

Note that your ETH collateral is sold for DAI in liquidation, so that the system is back to health. The system has no interest in keeping the collateral.

Implementing a liquidation engine is out of scope (see [here](https://github.com/yieldprotocol/vault-v2/blob/npm-0.10.0-rc0/contracts/Witch.sol) for context). For now, please, make it so that an user can be liquidated by the owner **only** if the value of the DAI they borrowed drops below the value of the collateral they deposited.